### ISSUE TITLE
Dynamic authentication and authorization via external WAMP client

### ISSUE THREAD

**Issue by ecorm** [darrenjs/wampcc#72]
As mentioned in my other issue, I'm considering this library as a replacement for Crossbar. I need to duplicate the dynamic authentication/authorization features, where Crossbar can call an RPC registered by an external client to authenticate the session or authorize a CALL/REGISTER/SUBSCRIBE/PUBLISH action.

The WAMP client app providing the authorizer/authenticator RPC would be running on the same machine as the wampcc router, and would be connected via TCP loopback. The latency would be very low, and I could cache the result on a per-user basis to avoid future round trips (with occasional clearing of the cache when user permissions are changed).

I see from the wampcc examples that custom authentication and authorization is indeed possible when the router is embedded within the server app.

TL;DR: **What I'd like to know is if it's safe (thread-wise and logic-wise) for the custom wampcc authentication/authorization handler to call an RPC (and receive the reply) before returning the `authorized` object.**

I know it would be cleaner to embed wampcc within my server app, but it's currently a large codebase that uses CppWAMP extensively (of which I'm the author, btw). Ideally, I would add router functionality to CppWAMP, but I'm currently looking for a quick fix to get rid of Crossbar.

**Comment by ecorm** [comment]
I was writing some code to test this, but it turns out that it's not possible to initiate an RPC from a `wamp_router` instance (unless I overlooked something).

I'm now at a crossroads, having to decide whether I should fork this project to add the missing functionality I need, or just roll-up my sleeves and add router functionality to CppWAMP (which I've always wanted to do anyway). The advantage of the latter is that I can embed the router in our backend server app without breaking the current API that it relies on.

In any case, I salute your fine work on this project. It's been quite easy to read your code to understand what's going on under the hood.

**Comment by ecorm** [comment]
I decided to go ahead and implement router functionality in CppWAMP. Please feel free to reopen this issue if needed.

**Comment by CanNuhlar** [comment]
Hi @ecorm, I'm trying to do the same here. I have implemented a custom router that can mimic(?) crossbar and make it compatible with autobahn-cpp. Are there any updates on your project CppWAMP?

And how far did you come with your fork of wampcc if you ever did?

Thanks!

**Comment by ecorm** [comment]
@CanNuhlar I ended up deciding to implement router functionality in CppWAMP. I recently had a basic router working in CppWAMP in my `router` branch, but I'm now in the in middle of a large refactoring to fix a design flaw in how servers accept new connections.

When I'm out of refactoring hell and get things compiling/working again, I can send you a commit hash from my `router` branch that you can evaluate, if that interests you.

Custom authenticator/authorizer objects can be passed to the CppWAMP router, and from there you can write your own code to forward authentication/authorization via RPC to an external client using what I call a "direct" client that talks directly to the router.

I've recently added Websocket support and am working on a basic HTTP server that can serve static files. I still don't support TLS connections, but I hope it will be easy enough to add. My goal is to get rid of python, crossbar, and NGINX dependencies so that I can implement a compact network appliance using an ARM-based single board computer using a minimal Linux image.

**Comment by ecorm** [comment]
@CanNuhlar If you like, you can try commit `3cfe2a44fd9fa2a6cc22ead930b8f7c7e95aa694` from the CppWAMP router branch. That was before my current server accept overhaul. The documentation on the router stuff is still spotty. There's a router example here: https://github.com/ecorm/cppwamp/blob/router/examples/router/main.cpp

If you have further question, please create a Q&A under the CppWAMP Discussions page: https://github.com/ecorm/cppwamp/discussions, so that we don't bother darrenjs.
